File: packages/cli/lib/privacyRedact.mjs
/**
 * Redact spent-leaf linkage fields from user-facing / shareable artifacts.
 * leafIndex may remain in local working stores for proving; never export in
 * shareable JSON / stdout (debug flags must not re-expose leaves).
 */

const LEAF_KEYS = new Set([
  "leafIndex",
  "leafIndices",
  "inLeafIndex",
  "paymentLeafIndex",
  "changeLeafIndex",
]);

/**
 * Deep-clone JSON-compatible value and strip leaf index fields.
 * Always redacts leaf fields; `opts.debug` / AP_PRIVACY_DEBUG are ignored.
 * @param {unknown} value
 * @param {{ debug?: boolean }} [opts]
 */
export function redactLeafIndexFields(value, _opts = {}) {
  return redact(value);
}

function redact(value) {
  if (Array.isArray(value)) return value.map(redact);
  if (value && typeof value === "object") {
    const out = {};
    for (const [k, v] of Object.entries(value)) {
      if (LEAF_KEYS.has(k)) continue;
      out[k] = redact(v);
    }
    return out;
  }
  return value;
}

/**
 * Minimal spend-note fields safe for sealed export (no leaf / depositor metadata).
 * @param {Record<string, unknown>} note
 */
export function minimalSpendNoteFields(note) {
  return {
    version: String(note.version),
    assetId: String(note.assetId),
    value: String(note.value),
    spendingKey: String(note.spendingKey),
    nullifierKey: String(note.nullifierKey),
    blinding: String(note.blinding),
    commitment: String(note.commitment),
  };
}

File: packages/python-client/absolute_privacy/privacy_redact.py
"""Strip leafIndex fields from shareable Python CLI/SDK outputs unless debug."""

from __future__ import annotations

from typing import Any

LEAF_KEYS = frozenset(
    {
        "leafIndex",
        "leafIndices",
        "inLeafIndex",
        "paymentLeafIndex",
        "changeLeafIndex",
    }
)


def redact_leaf_index_fields(value: Any, *, debug: bool = False) -> Any:
    if debug:
        return value
    if isinstance(value, list):
        return [redact_leaf_index_fields(v, debug=False) for v in value]
    if isinstance(value, dict):
        return {
            k: redact_leaf_index_fields(v, debug=False)
            for k, v in value.items()
            if k not in LEAF_KEYS
        }
    return value


def minimal_spend_note_fields(note: dict) -> dict:
    return {
        "version": str(note["version"]),
        "assetId": str(note["assetId"]),
        "value": str(note["value"]),
        "spendingKey": str(note["spendingKey"]),
        "nullifierKey": str(note["nullifierKey"]),
        "blinding": str(note["blinding"]),
        "commitment": str(note["commitment"]),
    }
